### ISSUE TITLE
Integer refinement type

### ISSUE THREAD

**Issue by njordhov** [clarity-lang/reference#13]
Issue #2 proposes to make uint a subtype of int, with the type checker and runtime system verifying that these are nonnegative integers. **Subtyping integers can be generalized by providing an integer [refinement type.](https://en.wikipedia.org/wiki/Refinement_type)**

Integers can be refined by specifying their range with a low and/or high boundary. A simple syntax [borrowed](https://www.cs.cmu.edu/Groups/AI/html/cltl/clm/node50.html) from Common Lisp is `(int low high)`. For example, here are some specific integer refinement types:

* _Digit_ is an integer from 0 to 9: `(int 0 9)`
* _Months_ are numbered from 1 to 12: `(int 1 12)`
* _Nonnegative_ integers (uint) go from 0 and up: `(int 0)`
* _Positive_ integers go from 1 and up: `(int 1)`

Integer refinement subtypes allows the type system to ensure that integers stay within the bounds. This can make Clarity code more succinct, opening for automatically verifying that integers stay in range rather than requiring defensive code to implement the boundary assertions. It also provides for more powerful type inference, where the possible boundaries of integer expressions can be determined during analytics.

Currently, integers are allocated 16 bytes (2^128) in storage, which can be wasteful. Knowing the range of integers can make the system pack integers to only use the actually required number of bytes.

**Comment by lgalabru** [comment]
I love this proposal

**Comment by kantai** [comment]
This is a great idea -- I think we'd need to specify behavior on the arithmetic functions a little more precisely before we'd know exactly how integer refinement types should work in Clarity.

For example:

```
(define-private (add-digits (a (int 0 9)) (b (int 0 9)))
  (+ a b))
```

What is the result of `(add-digits 5 5)`? Should it be `10` or a runtime error (or a checkable `(err ..)`)?

If the result should be`10` (which would be how most languages that support refinement types work), then our arithmetic functions would need to be dependently typed (with perhaps checked variants that error rather than increase the output type's range).

**Comment by njordhov** [comment]
> What is the result of (add-digits 5 5)? Should it be 10 or a runtime error 

It should **not** be a runtime error. Like [argued](https://github.com/blockstack/stacks-blockchain/pull/1683#issuecomment-647805500) for unsigned integer arithmetic, constraining arithmetic calculations to argument bounds leads to undesirable and surprising results. For example, this would underflow half the time:

```clojure
(define-read-only (difference (balance1 (int 0 100) (balance2 (int 0 100))
  (abs (- balance1 balance2)))
```

Instead, the return type of this function can be inferred to be `(int 0 100)`, constrained by the input types and the `abs` function. In cases where such an implicit constrain is insufficient, it may make sense to add an explicit constraint on the output, like [as-max-len?](https://docs.blockstack.org/core/smart/clarityref#as-max-len) does for sequences. Related: https://github.com/blockstack/stacks-blockchain/issues/1377

**Comment by psq** [comment]
how would you define negative integers without something like `MIN_INT`?  Or would you have to know what the min value is for 128 bits?

**Comment by njordhov** [comment]
> how would you define negative integers

Like `(int -100 0)`... ;-) but presume you mean _integers with an explicit upper bound but the default lower bound._ That's such a rare case, those that want it should provide the lower bound explicitly.  Besides, we may find it to be good practice anyway to specify explicit integer bounds.

**Comment by kantai** [comment]
> how would you define negative integers without something like MIN_INT? Or would you have to know what the min value is for 128 bits?

Yep -- it's almost certainly worthwhile to have constants like `max-int` or `min-int` for use cases like this.

**Comment by njordhov** [comment]
The integer refinement type could support **arbitrary precision for the upper and lower bounds.** That way, they can allow bigger integers like 256/512 bit as proposed in issue #8. The constants for `max-int` and `min-int` would then be the _default bounds_ not the actual limits for what can be specified.

**Comment by obycode** [comment]
I like this idea. I think the type inference could get tricky and should probably be separated from an initial implementation which uses this syntax to add 256- and 512-bit integers, and follows simple type-checking rules, such as those used currently by the arithmetic functions, i.e. argument types must match and the result will be of the same type.
